Build & deploy projects

elayabharath opened the pull request:
+ Added firebase config needed to deploy rules + hosting
+ Added build step to build all projects at once

elayabharath committed:
Added firebase rules to deploy

elayabharath committed:
Merge branch 'experimental-prod' into deploy-settings

elayabharath committed:
Added dev, build for multiple projects

elayabharath committed:
Remove unused path package

elayabharath committed:
Rename deploy folder to config

elayabharath committed:
Fix path to config

marianamarasoiu commented on config/firestore.rules:
```diff
@@ -0,0 +1,18 @@
+  match /databases/{database}/documents {
+    match /datasets/covid19-som/metadata/{document=**} {
+    	allow write: if
+      	request.auth.token.email_verified &&
```
probably not strictly necessary, but consider adding a null check to check specifically for unauthenticated access:
```suggestion
        request.auth != null &&
      	request.auth.token.email_verified &&
```

elayabharath replied:
Ah sure, will do

marianamarasoiu commented on config/firestore.rules:
```diff
@@ -0,0 +1,18 @@
+    }
+    match /{document=**} {
+      allow read: if
+      	request.auth.token.email_verified &&
```
same as above:
```suggestion
        request.auth != null &&
      	request.auth.token.email_verified &&
```

marianamarasoiu commented on config/project_config.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "covid19-som": {
+    "metadataPath": "datasets/covid19-som/metadata/chart-config",
+    "allowedEmailDomains": "lark.systems|africasvoices.org"
```
Not sure if I've mentioned this before, but you could consider using a list instead of a regex string - if I remember correctly, in the dart code you separate it into a list anyway, so that would just make it easier.

elayabharath replied:
Yes this is much easier to read, fixed.

marianamarasoiu commented on dev.dart:
```diff
@@ -0,0 +1,29 @@
+  var baseConfigFile = File('config/base_config.json');
+  var projConfigFile = File('config/project_config.json');
+  if (!await baseConfigFile.exists() || !await projConfigFile.exists()) {
+    throw ArgumentError('Config files not present');
```
Aim to make this error message a bit more comprehensive by testing the existence of both files separately (if the user gets this error, they wouldn't know which of the files is missing) and by listing the full path of the file so that the user would know where it should be added.

nit: "not present" as a phrase is a bit ambiguous - it could be interpreted as "not present in the argument list". Consider using something like "Could not find config file: <path>" or "Config file doesn't exist: <path>"

elayabharath replied:
That makes sense, separated out both of them to give unambiguous error messages. Fixed

marianamarasoiu commented on dev.dart:
```diff
@@ -0,0 +1,29 @@
+
+void main(List<String> arguments) async {
+  if (arguments == null) {
+    throw ArgumentError('dart dev.dart <proj-name-from-project_config.json>');
```
in the console this would look like:
`Invalid argument(s): dart dev.dart <proj-name-from-project_config.json>`
which doesn't read so well because "Invalid argument(s):" creates the expectation of "I'm going to get some info about what's wrong with the arguments, whereas the message is a usage message (the user will likely interpret it as such, but it will be a bit confusing at the beginning. A simple fix would be to just prefix the message with "Usage: "
```suggestion
    throw ArgumentError('Usage: dart dev.dart <proj-name-from-project_config.json>');
```

I'm also wondering if, instead of passing in the project name, and then checking it against some files in a default location, why not pass those files in as arguments to this script? That way, the user can be explicit as to which project configs files they would like to pass in, with the default ones being as an example (which you could give in the readme). What do you think?

marianamarasoiu replied:
or you could make the `config/base_config.json` and `config/project_config.json` as the default for running `dev.dart` without any args, and if the user is passing in other config files, it will use those?

elayabharath replied:
I thought of having one `base_config.json` and just one `project_config.json` where I will add configs that are project specific than a config file per project. That way I could avoid building logic to identify the config files and at the same time get an overview of all project configs at the same time. Would that be okay? 

I have changed the dev.dart such that if no arguments are provided, it simply takes the first project as the default one under `project_config.json`. Let me know if this is okay.

marianamarasoiu commented on config/base_config.json:
```diff
@@ -0,0 +1,12 @@
+  "measurementId": "G-S4J6TZ44LN",
+  "mapboxKey": "pk.eyJ1IjoiZWxheWFiaGFyYXRoIiwiYSI6ImNqMTRra3JubDAwMDYzM25uOXEwNDF3ZzIifQ.ntzLwhDXekJpMtyMpygz7g",
+  "mapboxStyleURL": "mapbox://styles/elayabharath/cjou0w3qa91ww2rp6gcafy3pt"
+}
```
would it be worth describing in the readme what the base_config.json and the project_config.json files are (and what the difference between them is)?

elayabharath replied:
Updated the readme with a description of both of them

marianamarasoiu commented on build.dart:
```diff
@@ -0,0 +1,40 @@
+    print('---------- ${proj} ----------');
+    print('${proj}: Building project ...');
+    var newConfig = {...baseConfig, ...projConfig};
+    var newFile = File('web/assets/firebase-constants.json');
```
it might make sense to rename this file from `firebase_constants.json` to just `constants.json` or `platform_constants.json` or something else, as it's currently not storing only firebase constants, but also mapbox constants

elayabharath replied:
Makes sense, renamed to `constants.json`

marianamarasoiu reviewed (approved):
thanks @elayabharath , a few comments and questions for you to consider

elayabharath committed:
Catch unauthenticated requests, before checking for email verification

Co-authored-by: Mariana Mărășoiu <mariana.marasoiu@gmail.com>

elayabharath committed:
Catch unauthenticated requests, before checking for email verification

Co-authored-by: Mariana Mărășoiu <mariana.marasoiu@gmail.com>

elayabharath committed:
Convert the config access emails to a list

elayabharath committed:
Update dev.dart

elayabharath committed:
Updated build & dev scripts, changed the constants file to remove firebase from the name

elayabharath committed:
Updated readme describing base config and project config

elayabharath commented:
Thanks @marianamarasoiu for the review. I have fixed most of them except the one related to `project_config.json` from above. Could you take a look if what I have changed on it is okay thing to do? Thank you!

marianamarasoiu commented on config/firestore.rules:
```diff
@@ -0,0 +1,20 @@
+    match /datasets/covid19-som/metadata/{document=**} {
+    	allow write: if
+        request.auth != null &&
+      	request.auth.token.email_verified &&
```
nit: there's some weird indentation going on, I've had this before, I think the firebase rules editor switches to using tabs sometimes. This should fix it:
```suggestion
        request.auth.token.email_verified &&
```

marianamarasoiu commented on config/firestore.rules:
```diff
@@ -0,0 +1,20 @@
+    match /{document=**} {
+      allow read: if
+        request.auth != null &&
+      	request.auth.token.email_verified &&
```
nit: indentation fix as above:
```suggestion
        request.auth.token.email_verified &&
```

marianamarasoiu reviewed (approved):
Ah, that makes sense, sorry, I think I misunderstood how you were using the project config. I think it's fine for now to use a single file to specify project-specific settings. What we might end up doing in the end is put those settings in firebase, and then when we want to add a new project, we can just specify it there rather than (also) committing to this repo (so basically being able to use the same firebase instance to host multiple projects, and adding the ability to select between them).

I'm currently working on the same thing for Nook (ie. multi-tenancy) - so let's talk more about how we can do it here when the Nook part is done.

in any case, LGTM for this, and thanks!

elayabharath commented:
Thanks @marianamarasoiu will look into the pattern you are using for nook and try to add similar deploy settings.
